File: prime/prime.go
package main

import "fmt"
import "sort"

func main() {
	for _, prime := range find_primes(100) {
		fmt.Printf("%d\n", prime)
	}
}

func find_primes(n int) []int {
	if n < 2 {
		return []int{}
	}

	candidates := []int{}
	primes := []int{2}
	for i := 3; i <= n; i += 2 {
		candidates = append(candidates, i)
	}

	for i := 0; len(candidates) > 0; i = i {
		primes = append(primes, candidates[i])
		for _, multiple := range multiples(candidates[i], n) {
			pos := sort.SearchInts(candidates, multiple)
			if pos != len(candidates) && candidates[pos] == multiple {
				candidates = append(candidates[:pos], candidates[pos+1:]...)
			}
		}
	}

	return primes
}

func multiples(n, max int) []int {
	multiples := []int{}
	for i := 1; i*n <= max; i++ {
		multiples = append(multiples, i*n)
	}
	return multiples
}

File: prime/prime.hs
multiples       :: Integer -> Integer -> [Integer]
multiples n max = [n, 2*n..max]

remove_from_list :: [Integer] -> [Integer] -> [Integer] -> [Integer]
remove_from_list target removal_array rv
    | null target || null removal_array =
        rv ++ target
    | head target > head removal_array =
        remove_from_list target (tail removal_array) rv
    | head target == head removal_array =
        remove_from_list (tail target) (tail removal_array) rv
    | otherwise =
        remove_from_list (tail target) removal_array (rv ++ [head target])

findprimes :: Integer -> [Integer]
findprimes n 
    | n < 2 = []
    | otherwise =
        2:(findprimes' [3,5..n] [])
    
findprimes' :: [Integer] -> [Integer] -> [Integer]
findprimes' candidates primes
    | null candidates = primes
    | otherwise =
        findprimes'
            (remove_from_list
                candidates
                (multiples (head candidates) (last candidates))
                [])
            (primes ++ [head candidates])

main = print (findprimes 100)

